torch/_inductor/triton_bundler.py

```diff
@@ -0,0 +1,188 @@
+                    path = os.path.join(entry.directory, entry.kernel_hash)
+                    if not os.path.exists(path):
+                        continue
+                    for filename in os.listdir(path):
+                        filepath = os.path.join(path, filename)
+                        try:
+                            assert os.path.isfile(filepath)
+                            with open(filepath, "rb") as file:
+                                artifacts.append(
+                                    TritonKernelArtifact(filename, file.read())
+                                )
+                            counters["inductor"]["triton_bundler_save_kernel"] += 1
+                        except Exception:
+                            log.debug("failed to collect triton kernel", exc_info=True)
+                    if artifacts:
+                        result.append(
+                            TritonKernelArtifacts(
+                                entry.kernel_hash,
+                                entry.compile_time_ns,
+                                entry.device,
+                                artifacts,
+                            )
+                        )
+                        total_compile_time_ns += entry.compile_time_ns
+                cls._entries = None
+                num_kernels = len(result)
+                return result, TritonBundlerMetadata(num_kernels, total_compile_time_ns)
+            return [], None
+
+    @staticmethod
+    def write(bundle: List[TritonKernelArtifacts]) -> Optional[TritonBundlerMetadata]:
+        if not TritonBundler.is_enabled():
+            return None
+
+        with dynamo_timed(
+            key="TritonBundler::write", phase_name="TritonBundler", fwd_only=False
+        ):
+            num_kernels = 0
+            total_compile_time_ns = 0
+            basedir = os.getenv("TRITON_CACHE_DIR")
+            custom_dir = True
+            if basedir is None:
+                custom_dir = False
+                basedir = os.path.join(cache_dir(), "triton")
+
+            for artifacts in bundle:
+                directory = os.path.join(basedir, artifacts.kernel_hash)
+                if not custom_dir:
+                    # When we do not use a custom triton cache dir, we separate kernels
+                    # by their device
+                    directory = os.path.join(directory, str(artifacts.device))
+                Path(directory).mkdir(parents=True, exist_ok=True)
+
+                with tempfile.TemporaryDirectory() as tmp_dir:
+                    for artifact in artifacts.artifacts:
+                        filepath = os.path.join(tmp_dir, artifact.filename)
+                        with open(filepath, "wb") as file:
+                            file.write(artifact.payload)
+                        counters["inductor"]["triton_bundler_write_kernel"] += 1
+                    shutil.copytree(tmp_dir, directory, dirs_exist_ok=True)
```

Comment:
... why a copytree?

Reply:
is there a better way to copy a folder?

Reply:
Naively I would have expected you to write the files directly into the locations they were expected to be (or, if atomicity is needed, rename them in), not sure why we write it one place and then do a copy

Reply:
That's what I originally did, but Colin asked me to change it since what if we fail midway writing files, we want all or nothing behavior

Reply:
You can still do that though, you want to do an atomic file rename. In fact, copytree is implemented under the hood by writing out 16KB chunks https://stackoverflow.com/questions/20873723/is-pythons-shutil-copyfile-atomic so it's not even atomic!

Reply:
@c00w ?